File: data/figures/chapter05/figure_5_21/plot_5_21.m
function plot_5_21(outputFile)
if nargin < 1, outputFile=fullfile(fileparts(mfilename('fullpath')),'plot_5_21.png'); end
T=loadSiblingData('figure_5_21.csv');

O=T(textCol(T,'record_type')=="observed",:); R=T(textCol(T,'record_type')=="reference",:); feasible=logicalCol(O,'Feasible'); pf=logicalCol(O,'Pareto'); fig=figure('Visible','off','Color','w','Position',[100 100 720 1000]); tl=tiledlayout(fig,3,1,'Padding','compact','TileSpacing','compact'); ax=nexttile(tl); scatter(ax,numericSubset(O,'TaperRatio',feasible),1000*numericSubset(O,'CDitrim',feasible),34,numericSubset(O,'AR',feasible),'filled'); hold(ax,'on'); scatter(ax,numericSubset(O,'TaperRatio',~feasible),1000*numericSubset(O,'CDitrim',~feasible),30,'x','MarkerEdgeColor',[.6 .6 .6]); scatter(ax,numericSubset(O,'TaperRatio',pf),1000*numericSubset(O,'CDitrim',pf),65,'o','MarkerFaceColor','none','MarkerEdgeColor','k'); xlabel(ax,'Taper ratio'); ylabel(ax,'1000 C_{D_i,trim}');
ax=nexttile(tl); ar=numericCol(R,'AR'); taper=numericCol(R,'TaperRatio'); for value=[6 9 12], m=abs(ar-value)<1e-12; [x,ord]=sort(taper(m)); y=numericCol(R,'RigidSpanEfficiency'); y=y(m); plot(ax,x,y(ord),'LineWidth',1.1,'DisplayName',sprintf('Rigid LLT, AR=%g',value)); hold(ax,'on'); end; scatter(ax,numericSubset(O,'TaperRatio',feasible),numericSubset(O,'TorsionCorrectedE',feasible),28,numericSubset(O,'AR',feasible),'filled'); scatter(ax,numericSubset(O,'TaperRatio',~feasible),numericSubset(O,'TorsionCorrectedE',~feasible),28,'x','MarkerEdgeColor',[.6 .6 .6]); xlabel(ax,'Taper ratio'); ylabel(ax,'Span efficiency'); legend(ax,'Location','best');
ax=nexttile(tl); ut=unique(taper); lo=nan(size(ut)); hi=lo; pen=numericCol(R,'TaperOnlyCDiPenaltyPct'); for i=1:numel(ut), vals=pen(abs(taper-ut(i))<1e-12); lo(i)=min(vals); hi(i)=max(vals); end; fill(ax,[ut;flipud(ut)],[lo;flipud(hi)],[.85 .92 .97],'EdgeColor','none','DisplayName','Rigid LLT envelope'); hold(ax,'on'); m=abs(ar-9)<1e-12; [x,ord]=sort(taper(m)); y=pen(m); plot(ax,x,y(ord),'LineWidth',1.2,'DisplayName','Rigid LLT, AR=9'); scatter(ax,numericSubset(O,'TaperRatio',feasible),numericSubset(O,'TaperOnlyCDiPenaltyPct',feasible),28,numericSubset(O,'AR',feasible),'filled'); scatter(ax,numericSubset(O,'TaperRatio',~feasible),numericSubset(O,'TaperOnlyCDiPenaltyPct',~feasible),28,'x','MarkerEdgeColor',[.6 .6 .6]); xlabel(ax,'Taper ratio'); ylabel(ax,'Taper-only C_{D_i} penalty (%)'); legend(ax,'Location','best'); colormap(fig,'parula'); title(tl,'Taper-ratio theory and observed-result consistency checks'); finishPlot(fig,outputFile);
end

function T = loadSiblingData(csvName)
here = fileparts(mfilename('fullpath'));
T = readtable(fullfile(here, csvName), 'VariableNamingRule', 'preserve');
end
function v = numericCol(T, name)
x = T.(name);
if isnumeric(x) || islogical(x)
    v = double(x);
else
    v = str2double(string(x));
end
v = v(:);
end
function v = numericSubset(T, name, index)
v = numericCol(T,name); v = v(index);
end
function s = textCol(T, name)
s = string(T.(name)); s = s(:);
end
function tf = logicalCol(T, name)
x = T.(name);
if islogical(x)
    tf = x;
elseif isnumeric(x)
    tf = x ~= 0;
else
    tf = ismember(lower(strtrim(string(x))), ["1","true","yes"]);
end
tf = tf(:);
end
function [xv, yv, Z] = gridData(T, xName, yName, zName)
x = numericCol(T,xName); y = numericCol(T,yName); z = numericCol(T,zName);
valid = isfinite(x) & isfinite(y) & isfinite(z); x=x(valid); y=y(valid); z=z(valid);
xv = unique(x,'sorted'); yv = unique(y,'sorted');
[~,ix] = ismember(x,xv); [~,iy] = ismember(y,yv);
Z = nan(numel(yv),numel(xv)); Z(sub2ind(size(Z),iy,ix)) = z;
end
function plotObservations(ax,x,y,feasible)
scatter(ax,x(~feasible),y(~feasible),28,[0.70 0.09 0.17],'x','LineWidth',1.0,'DisplayName','Infeasible'); hold(ax,'on');
scatter(ax,x(feasible),y(feasible),30,'o','MarkerFaceColor','w','MarkerEdgeColor',[0.09 0.41 0.67],'LineWidth',0.9,'DisplayName','Feasible');
end
function plotPareto(ax,T,xName,yName,feasName,pfName,caseName)
x=numericCol(T,xName); y=numericCol(T,yName); feasible=logicalCol(T,feasName); pf=logicalCol(T,pfName);
plotObservations(ax,x,y,feasible); [sx,ord]=sort(x(pf)); sy=y(pf); sy=sy(ord); plot(ax,sx,sy,'-','Color',[0.13 0.13 0.13],'LineWidth',0.9,'HandleVisibility','off');
scatter(ax,x(pf),y(pf),36,[0.13 0.13 0.13],'filled','DisplayName','Observed Pareto');
cases=numericCol(T,caseName); ids=find(pf); for k=1:numel(ids), text(ax,x(ids(k)),y(ids(k)),sprintf(' %d',round(cases(ids(k)))),'FontSize',7); end
end
function h = plotScore(ax,G,P,xName,yName,scoreName,pxName,pyName,feasName,pfName)
[xv,yv,Z]=gridData(G,xName,yName,scoreName); h=imagesc(ax,xv,yv,Z); set(ax,'YDir','normal'); hold(ax,'on');
if min(Z,[],'all','omitnan') <= 0.5 && max(Z,[],'all','omitnan') >= 0.5, contour(ax,xv,yv,Z,[0.5 0.5],'w','LineWidth',1.0); end
xp=numericCol(P,pxName); yp=numericCol(P,pyName); feasible=logicalCol(P,feasName); pf=logicalCol(P,pfName); plotObservations(ax,xp,yp,feasible); scatter(ax,xp(pf),yp(pf),55,'o','MarkerFaceColor','none','MarkerEdgeColor',[0.13 0.13 0.13],'LineWidth',1.1,'DisplayName','Observed Pareto'); axis(ax,'square');
end
function finishPlot(fig, outputFile)
folder=fileparts(outputFile); if ~isempty(folder) && ~isfolder(folder), mkdir(folder); end
exportgraphics(fig,outputFile,'Resolution',180,'BackgroundColor','white'); close(fig);
end

File: data/figures/chapter05/figure_5_21/plot_5_21.py
from __future__ import annotations
import argparse
import csv
import math
from pathlib import Path
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D
HERE = Path(__file__).resolve().parent
CSV_FILE = HERE / "figure_5_21.csv"
INK="#202124"; BLUE="#1769aa"; ORANGE="#e66101"; RED="#b2182b"; PURPLE="#7b3294"; GREY="#9aa0a6"; LIGHT="#d9e8f5"
plt.rcParams.update({"font.family":"Times New Roman","font.size":9,"axes.titlesize":11,"axes.labelsize":9,"legend.fontsize":8})
def load_rows():
    with CSV_FILE.open("r", encoding="utf-8-sig", newline="") as handle:
        return list(csv.DictReader(handle))
def select(rows, kind):
    return [r for r in rows if r.get("record_type", "") == kind]
def num(rows, name):
    out=[]
    for row in rows:
        value=row.get(name, "")
        out.append(float(value) if value not in ("", "nan", "NaN") else np.nan)
    return np.asarray(out, dtype=float)
def text(rows, name):
    return np.asarray([row.get(name, "") for row in rows], dtype=str)
def flag(rows, name):
    values=text(rows,name)
    return np.asarray([v.strip().lower() in {"1","true","yes"} for v in values], dtype=bool)
def gridify(rows, xname, yname, zname):
    x=num(rows,xname); y=num(rows,yname); z=num(rows,zname)
    xs=np.unique(x); ys=np.unique(y); image=np.full((len(ys),len(xs)),np.nan)
    ix=np.searchsorted(xs,x); iy=np.searchsorted(ys,y); image[iy,ix]=z
    return xs,ys,image
def observations(ax,x,y,feasible):
    ax.scatter(x[~feasible],y[~feasible],marker="x",c=RED,s=24,label="Infeasible",zorder=4)
    ax.scatter(x[feasible],y[feasible],facecolors="white",edgecolors=BLUE,s=25,label="Feasible",zorder=4)
def pareto_panel(ax,rows,xname,yname,feasname,pfname,casename):
    x=num(rows,xname); y=num(rows,yname); feas=flag(rows,feasname); pf=flag(rows,pfname)
    observations(ax,x,y,feas)
    ax.scatter(x[feas & ~pf],y[feas & ~pf],facecolors="white",edgecolors=BLUE,s=25)
    order=np.flatnonzero(pf)[np.argsort(x[pf])]
    ax.plot(x[order],y[order],color=INK,lw=0.9,zorder=2)
    ax.scatter(x[pf],y[pf],c=INK,s=30,label="Observed Pareto",zorder=5)
    cases=num(rows,casename)
    for i in np.flatnonzero(pf): ax.annotate(str(int(cases[i])),(x[i],y[i]),xytext=(3,3),textcoords="offset points",fontsize=6.5)
def score_panel(ax,grid,points,xname,yname,score,px,py,feasname,pfname,title):
    xs,ys,z=gridify(grid,xname,yname,score)
    image=ax.pcolormesh(xs,ys,z,shading="auto",cmap="cividis",vmin=0,vmax=1)
    if np.nanmin(z)<=0.5<=np.nanmax(z): ax.contour(xs,ys,z,levels=[0.5],colors="white",linewidths=1.0)
    xp=num(points,px); yp=num(points,py); feas=flag(points,feasname); pf=flag(points,pfname)
    observations(ax,xp,yp,feas); ax.scatter(xp[pf],yp[pf],s=45,facecolors="none",edgecolors=INK,lw=1.0,zorder=5)
    ax.set_box_aspect(1); ax.set_title(title)
    return image
def parse_output():
    parser=argparse.ArgumentParser(); parser.add_argument("--output",type=Path,default=HERE/"plot_5_21.png")
    return parser.parse_args().output
def finish(fig):
    out=parse_output(); out.parent.mkdir(parents=True,exist_ok=True); fig.savefig(out,dpi=180,bbox_inches="tight"); plt.close(fig); print(out)

def main():
    d=load_rows(); obs=select(d,"observed"); ref=select(d,"reference"); ar=num(ref,"AR"); taper=num(ref,"TaperRatio"); chosen=np.asarray([6.,9.,12.]); fig,axes=plt.subplots(3,1,figsize=(7.2,10),constrained_layout=True)
    feas=flag(obs,"Feasible"); axes[0].scatter(num(obs,"TaperRatio")[feas],1000*num(obs,"CDitrim")[feas],c=num(obs,"AR")[feas],cmap="viridis",s=32,label="Feasible"); axes[0].scatter(num(obs,"TaperRatio")[~feas],1000*num(obs,"CDitrim")[~feas],marker="x",c=GREY,s=30,label="Infeasible"); pf=flag(obs,"Pareto"); axes[0].scatter(num(obs,"TaperRatio")[pf],1000*num(obs,"CDitrim")[pf],s=60,facecolors="none",edgecolors=INK); axes[0].set_ylabel("Trim induced drag, 1000 CDi")
    for value in chosen:
        m=np.isclose(ar,value); order=np.argsort(taper[m]); axes[1].plot(taper[m][order],num(ref,"RigidSpanEfficiency")[m][order],label=f"Rigid LLT, AR={value:g}")
    axes[1].scatter(num(obs,"TaperRatio")[feas],num(obs,"TorsionCorrectedE")[feas],c=num(obs,"AR")[feas],cmap="viridis",s=30); axes[1].scatter(num(obs,"TaperRatio")[~feas],num(obs,"TorsionCorrectedE")[~feas],marker="x",c=GREY,s=28); axes[1].set_ylabel("Span efficiency"); axes[1].legend(frameon=False,ncol=3)
    ut=np.unique(taper); lo=[]; hi=[]
    for t in ut:
        vals=num(ref,"TaperOnlyCDiPenaltyPct")[np.isclose(taper,t)]; lo.append(vals.min()); hi.append(vals.max())
    axes[2].fill_between(ut,lo,hi,color=LIGHT,label="Rigid LLT envelope, AR=6-12"); m=np.isclose(ar,9); order=np.argsort(taper[m]); axes[2].plot(taper[m][order],num(ref,"TaperOnlyCDiPenaltyPct")[m][order],color=BLUE,label="Rigid LLT, AR=9"); axes[2].scatter(num(obs,"TaperRatio")[feas],num(obs,"TaperOnlyCDiPenaltyPct")[feas],c=num(obs,"AR")[feas],cmap="viridis",s=30); axes[2].scatter(num(obs,"TaperRatio")[~feas],num(obs,"TaperOnlyCDiPenaltyPct")[~feas],marker="x",c=GREY,s=28); axes[2].set_ylabel("Taper-only CDi penalty (%)"); axes[2].legend(frameon=False)
    for ax in axes: ax.set_xlabel("Taper ratio")
    fig.suptitle("Taper-ratio theory and observed-result consistency checks"); finish(fig)

if __name__ == "__main__":
    main()
